core/src/main/java/com/linecorp/armeria/common/multipart/ConcatArrayStreamMessage.java

```diff
@@ -0,0 +1,231 @@
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+        parent = new ConcatArraySubscriber<>(subscriber, sources, executor, options);
+        subscriber.onSubscribe(parent);
+        parent.nextSource();
+    }
+
+    @Override
+    public void abort() {
+        for (StreamMessage<? extends T> source : sources) {
+            source.abort();
+        }
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        for (StreamMessage<? extends T> source : sources) {
+            source.abort(cause);
+        }
+    }
+
+    private static final class ConcatArraySubscriber<T> extends SubscriptionArbiter implements Subscriber<T> {
+
+        private static final long serialVersionUID = -9184116713095894096L;
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<ConcatArraySubscriber> cancelledUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(ConcatArraySubscriber.class, "cancelled");
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<ConcatArraySubscriber> wipUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(ConcatArraySubscriber.class, "wip");
+
+        private Subscriber<? super T> downstream;
+        private final StreamMessage<? extends T>[] sources;
+        private final EventExecutor executor;
+        private final SubscriptionOption[] options;
+
+        private long produced;
+
+        private volatile int index;
+        private volatile int wip;
+        private volatile int cancelled;
+
+        ConcatArraySubscriber(Subscriber<? super T> downstream, StreamMessage<? extends T>[] sources,
+                              EventExecutor executor, SubscriptionOption... options) {
+            this.downstream = downstream;
+            this.sources = sources;
+            this.executor = executor;
+            this.options = options;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            if (cancelled != 0) {
+                subscription.cancel();
+            } else {
+                setSubscription(subscription);
+            }
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            produced++;
+            downstream.onNext(item);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            downstream.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            final long produced = this.produced;
+            if (produced != 0L) {
+                this.produced = 0L;
+                produced(produced);
+            }
+            nextSource();
+        }
+
+        void nextSource() {
+            if (wipUpdater.getAndIncrement(this) == 0) {
```

Comment:
> Is this for the case where nextSource() is called twice concurrently?

I don't think so. The `nextSource()` should be called only when a source is completed.

Reply:
If so, I think we don't need `wipUpdater`. Do we?

Reply:
`wipUpdater` and `volatile` keyword from `wip` could be removed. However, we still need `wip` counter because of an empty publisher that will call `onComplete()` while subscribing.